File: feedback_api.py
#!/usr/bin/env python3
"""
피드백 저장용 Flask API 서버
Streamlit과 별도로 실행되어 JavaScript에서 직접 피드백을 전송받습니다.
"""

from flask import Flask, request, jsonify
from flask_cors import CORS
from database import ModelDatabase
import json
import logging

# Flask 앱 생성
app = Flask(__name__)
CORS(app)  # CORS 허용

# 로깅 설정
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

@app.route('/save_feedback', methods=['POST', 'OPTIONS'])
def save_feedback():
    """피드백 저장 API"""
    
    # CORS preflight 요청 처리
    if request.method == 'OPTIONS':
        response = jsonify({'status': 'ok'})
        response.headers.add('Access-Control-Allow-Origin', '*')
        response.headers.add('Access-Control-Allow-Headers', 'Content-Type')
        response.headers.add('Access-Control-Allow-Methods', 'POST')
        return response
    
    try:
        # JSON 데이터 받기
        feedback_data = request.get_json()
        
        if not feedback_data:
            return jsonify({'error': 'No data provided'}), 400
        
        logger.info(f"📝 피드백 데이터 수신: {feedback_data}")
        
        # 필수 필드 확인
        required_fields = ['model_id', 'x', 'y', 'z', 'screen_x', 'screen_y', 'comment']
        for field in required_fields:
            if field not in feedback_data:
                return jsonify({'error': f'Missing field: {field}'}), 400
        
        # 데이터베이스에 저장
        db = ModelDatabase()
        feedback_id = db.add_feedback(
            model_id=feedback_data['model_id'],
            x=float(feedback_data['x']),
            y=float(feedback_data['y']),
            z=float(feedback_data['z']),
            screen_x=float(feedback_data['screen_x']),
            screen_y=float(feedback_data['screen_y']),
            comment=str(feedback_data['comment']),
            feedback_type=feedback_data.get('feedback_type', 'point')
        )
        
        if feedback_id:
            logger.info(f"✅ 피드백 저장 성공 - ID: {feedback_id}")
            return jsonify({
                'success': True, 
                'feedback_id': feedback_id,
                'message': '피드백이 성공적으로 저장되었습니다.'
            }), 200
        else:
            logger.error("❌ 피드백 저장 실패")
            return jsonify({'error': 'Failed to save feedback'}), 500
            
    except Exception as e:
        logger.error(f"❌ 피드백 저장 오류: {str(e)}")
        return jsonify({'error': str(e)}), 500

@app.route('/health', methods=['GET'])
def health_check():
    """서버 상태 확인"""
    return jsonify({'status': 'healthy', 'message': 'Feedback API server is running'})

if __name__ == '__main__':
    print("🚀 피드백 API 서버 시작 중...")
    print("📡 엔드포인트: http://localhost:5001/save_feedback")
    print("🔍 상태 확인: http://localhost:5001/health")
    app.run(host='0.0.0.0', port=5001, debug=True)

File: web_server_scripts/feedback_api.php
<?php
/**
 * 피드백 API
 * 피드백 저장, 조회, 수정, 삭제 기능
 */

// CORS 헤더 설정
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');
header('Content-Type: application/json');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    exit(0);
}

try {
    // 데이터베이스 연결
    $db = new PDO('sqlite:streamlit_3d.db');
    $db->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    
    $method = $_SERVER['REQUEST_METHOD'];
    $action = $_GET['action'] ?? '';
    
    switch ($method) {
        case 'POST':
            if ($action === 'save') {
                // 피드백 저장
                $input = json_decode(file_get_contents('php://input'), true);
                if (!$input) {
                    throw new Exception('Invalid JSON data');
                }
                
                $required = ['model_id', 'x', 'y', 'z', 'screen_x', 'screen_y', 'comment'];
                foreach ($required as $field) {
                    if (!isset($input[$field])) {
                        throw new Exception("Missing field: $field");
                    }
                }
                
                $stmt = $db->prepare("
                    INSERT INTO feedbacks 
                    (model_id, x, y, z, screen_x, screen_y, comment, feedback_type, status, created_at)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'pending', datetime('now'))
                ");
                
                $stmt->execute([
                    $input['model_id'],
                    floatval($input['x']),
                    floatval($input['y']),
                    floatval($input['z']),
                    floatval($input['screen_x']),
                    floatval($input['screen_y']),
                    $input['comment'],
                    $input['feedback_type'] ?? 'point'
                ]);
                
                $feedback_id = $db->lastInsertId();
                
                echo json_encode([
                    'success' => true,
                    'feedback_id' => $feedback_id,
                    'message' => 'Feedback saved successfully'
                ]);
                
            } else {
                throw new Exception('Invalid action for POST');
            }
            break;
            
        case 'GET':
            if ($action === 'list') {
                // 특정 모델의 피드백 목록 조회
                $model_id = $_GET['model_id'] ?? '';
                if (!$model_id) {
                    throw new Exception('model_id is required');
                }
                
                $stmt = $db->prepare("
                    SELECT id, model_id, x, y, z, screen_x, screen_y, comment, 
                           feedback_type, status, created_at
                    FROM feedbacks 
                    WHERE model_id = ?
                    ORDER BY created_at DESC
                ");
                $stmt->execute([$model_id]);
                $feedbacks = $stmt->fetchAll(PDO::FETCH_ASSOC);
                
                echo json_encode([
                    'success' => true,
                    'feedbacks' => $feedbacks,
                    'count' => count($feedbacks)
                ]);
                
            } else {
                throw new Exception('Invalid action for GET');
            }
            break;
            
        case 'PUT':
            if ($action === 'update_status') {
                // 피드백 상태 업데이트
                $input = json_decode(file_get_contents('php://input'), true);
                if (!$input || !isset($input['id']) || !isset($input['status'])) {
                    throw new Exception('Missing id or status');
                }
                
                $stmt = $db->prepare("
                    UPDATE feedbacks 
                    SET status = ? 
                    WHERE id = ?
                ");
                $stmt->execute([$input['status'], $input['id']]);
                
                echo json_encode([
                    'success' => true,
                    'message' => 'Feedback status updated'
                ]);
                
            } else {
                throw new Exception('Invalid action for PUT');
            }
            break;
            
        case 'DELETE':
            // 피드백 삭제
            $feedback_id = $_GET['id'] ?? '';
            if (!$feedback_id) {
                throw new Exception('feedback id is required');
            }
            
            $stmt = $db->prepare("DELETE FROM feedbacks WHERE id = ?");
            $stmt->execute([$feedback_id]);
            
            echo json_encode([
                'success' => true,
                'message' => 'Feedback deleted successfully'
            ]);
            break;
            
        default:
            throw new Exception('Method not allowed');
    }
    
} catch (PDOException $e) {
    http_response_code(500);
    echo json_encode([
        'success' => false,
        'error' => 'Database error: ' . $e->getMessage()
    ]);
} catch (Exception $e) {
    http_response_code(500);
    echo json_encode([
        'success' => false,
        'error' => $e->getMessage()
    ]);
}
?>
